core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -0,0 +1,430 @@
+
+          info(s"Finished loading ${loadedTransactions.size} transaction metadata from $topicPartition in ${time.milliseconds() - startMs} milliseconds")
+        }
+    }
+  }
+
+  /**
+   * When this broker becomes a leader for a transaction log partition, load this partition and
+   * populate the transaction metadata cache with the transactional ids.
+   */
+  def loadTransactionsForPartition(partition: Int, coordinatorEpoch: Int) {
+    validateTransactionTopicPartitionCountIsStable()
+
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
+
+    inLock(stateLock) {
+      ownedPartitions.put(partition, coordinatorEpoch)
+      loadingPartitions.add(partition)
+    }
+
+    def loadTransactions() {
+      info(s"Loading transaction metadata from $topicPartition")
+      try {
+        loadTransactionMetadata(topicPartition)
+      } catch {
+        case t: Throwable => error(s"Error loading transactions from transaction log $topicPartition", t)
+      } finally {
+        inLock(stateLock) {
+          loadingPartitions.remove(partition)
+        }
+      }
+    }
+
+    scheduler.schedule(topicPartition.toString, loadTransactions)
+  }
+
+  /**
+   * When this broker becomes a follower for a transaction log partition, clear out the cache for corresponding transactional ids
+   * that belong to that partition.
+   */
+  def removeTransactionsForPartition(partition: Int) {
+    validateTransactionTopicPartitionCountIsStable()
+
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partition)
+
+    inLock(stateLock) {
+      ownedPartitions.remove(partition)
+    }
+
+    def removeTransactions() {
+      var numTxnsRemoved = 0
+
+      inLock(stateLock) {
+        for (transactionalId <- transactionMetadataCache.keys) {
+          if (partitionFor(transactionalId) == partition) {
+            // we do not need to worry about whether the transactional id has any ongoing transaction or not since
+            // the new leader will handle it
+            transactionMetadataCache.remove(transactionalId)
+            numTxnsRemoved += 1
+          }
+        }
+
+        if (numTxnsRemoved > 0)
+          info(s"Removed $numTxnsRemoved cached transaction metadata for $topicPartition on follower transition")
+      }
+    }
+
+    scheduler.schedule(topicPartition.toString, removeTransactions)
+  }
+
+  private def validateTransactionTopicPartitionCountIsStable(): Unit = {
+    val curTransactionTopicPartitionCount = getTransactionTopicPartitionCount
+    if (transactionTopicPartitionCount != curTransactionTopicPartitionCount)
+      throw new KafkaException(s"Transaction topic number of partitions has changed from $transactionTopicPartitionCount to $curTransactionTopicPartitionCount")
+  }
+
+  def appendTransactionToLog(transactionalId: String,
+                             txnMetadata: TransactionMetadata,
+                             responseCallback: Errors => Unit) {
```

Comment:
This is about the inter-broker protocol version. Details are here:

http://kafka.apache.org/documentation/#upgrade_10_1

Maybe just leave a TODO marker and I can address it in a follow-up PR, so we would not drag too long for this one?